src/matching.rs

```diff
@@ -0,0 +1,585 @@
+use std::collections::VecDeque;
+use std::hash::Hash;
+
+use super::visit::{
+    EdgeRef, GraphBase, IntoEdges, IntoNeighbors, IntoNodeIdentifiers, NodeCompactIndexable,
+    NodeIndexable, VisitMap, Visitable,
+};
+
+/// Computed
+/// [*matching*](https://en.wikipedia.org/wiki/Matching_(graph_theory)#Definitions)
+/// of the graph.
+pub struct Matching<G: GraphBase> {
+    graph: G,
+    mate: Vec<G::NodeId>,
+    n_edges: usize,
+}
+
+impl<G> Matching<G>
+where
+    G: NodeCompactIndexable,
+{
+    fn new(graph: G, mate: Vec<G::NodeId>, n_edges: usize) -> Self {
+        Self {
+            graph,
+            mate,
+            n_edges,
+        }
+    }
+
+    /// Gets the matched counterpart of given node, if there is any.
+    ///
+    /// Returns `None` if the node is not matched.
+    ///
+    /// # Panics
```

Comment:
Done. One question regarding panics: After I added support for stable graph, the sentence "Panics if the node does not exist" is not precisely true anymore. It will panic only if given index is out of bounds, but not if it corresponds to a hole (i.e., removed node) in `mate` array. Should I change the sentence or is it acceptable as it is now?

Reply:
Nice catch. The best option there would be to specify when does the panic happen (i.e. if bigger than node_bound).

I just did a quick test changing the index operation for
```rust
 let &mate = self.mate.get(self.graph.to_index(node))?;
```
and it actually seems faster, because it removes a possible execution branch for the panic (in some quick benchmark querrying the mate of each vector). That's another option too.

Reply:
This is perfect! So I got rid of the panics, `None` or `false` are returned instead in those cases. I also put it explicitly in the documentation.